File: core/shared/src/main/scala/com/lynbrookrobotics/potassium/units/Histogram.java
package com.lynbrookrobotics.potassium.units;

import java.util.function.DoubleConsumer;

/**
 * Created by the-magical-llamicorn on 4/19/17.
 */
public class Histogram implements DoubleConsumer {

    protected final long[] histogram;
    public final double min, max;
    public final double interval;
    public final int bins;

    public Histogram(final double min, final double max, final int bins) {
        this.min = min;
        this.max = max;
        this.bins = bins;
        interval = (max - min) / bins;
        histogram = new long[bins + 2]; // bins and "less than min" and "greater than max"
    }

    public void accept(final double value) {
        if (value < min) histogram[0]++;
        else if (value > max) histogram[bins + 1]++;
        else histogram[1 + (int) ((value - min) / interval)]++;
    }

    public String toString() {
        final StringBuilder sb = new StringBuilder();
        sb.append("<");
        sb.append(min);
        sb.append(" : ");
        sb.append(histogram[0]);
        sb.append('\n');

        for (int i = 1; i < bins + 1; i++) {
            sb.append((float) (min + (interval * (i - 1))));
            sb.append(" to ");
            sb.append((float) (min + (interval * i)));
            sb.append(" : ");
            sb.append(histogram[i]);
            sb.append('\n');
        }

        sb.append(">");
        sb.append(max);
        sb.append(" : ");
        sb.append(histogram[bins+1]);
        sb.append('\n');

        return sb.toString();
    }
}

File: core/shared/src/main/scala/com/lynbrookrobotics/potassium/units/Histogram.scala
package com.lynbrookrobotics.potassium.units

import squants.time.Time


import com.team846.frc2015.logging.AsyncLogger

/**
  * Created by the-magical-llamicorn on 4/19/17.
  */
class Histogram(min: Time, max: Time, binCount: Int) {
  val interval = (max - min) / binCount

  // bins and "less than min" and "greater than max"
  val bins = new Array[Long](binCount + 2)

  def apply(value: Time) {
    if (value < min) {
      bins(0) += 1
    } else if (value > max) {
      bins(binCount + 1) += 1
    } else {
      bins(1 + ((value - min) / interval).toInt) += 1
    }
  }

  def printStatus(): Unit =  {
    AsyncLogger.info(s"< $min : ${bins(0)} \n")

    for  (i <- 1 until binCount) {
        AsyncLogger.info(s"${bins(i)} to ${min + (interval * (i -  1))} : " +
                s"${min + (interval * i)}")
    }

    AsyncLogger.info(s"> $max : ${bins(binCount + 1)}")
  }
}